include/unifex/indexed_for.hpp

```diff
@@ -0,0 +1,169 @@
+
+#include <unifex/config.hpp>
+#include <unifex/receiver_concepts.hpp>
+#include <unifex/sender_concepts.hpp>
+#include <unifex/type_traits.hpp>
+#include <unifex/blocking.hpp>
+#include <unifex/get_stop_token.hpp>
+#include <unifex/async_trace.hpp>
+
+#include <functional>
+#include <type_traits>
+
+namespace execution {
+class sequenced_policy;
+class parallel_policy;
+}
+
+namespace unifex {
+
+template <typename Predecessor, typename Policy, typename Range, typename Func>
+struct indexed_for_sender {
+  UNIFEX_NO_UNIQUE_ADDRESS Predecessor pred_;
+  UNIFEX_NO_UNIQUE_ADDRESS Policy policy_;
+  UNIFEX_NO_UNIQUE_ADDRESS Range range_;
+  UNIFEX_NO_UNIQUE_ADDRESS Func func_;
+
+  template <template <typename...> class Tuple>
+  struct indexed_for_result {
+   public:
+    template <typename... Args>
+    using apply = Tuple<Args...>;
+  };
+
+  template <typename... Args>
+  using is_overload_noexcept = std::bool_constant<noexcept(
+      std::invoke(std::declval<Func>(), std::declval<Args>()...))>;
+
+  template <template <typename...> class Variant>
+  struct calculate_errors {
+   public:
+    template <typename... Errors>
+    using apply = std::conditional_t<
+        Predecessor::
+            template value_types<std::conjunction, is_overload_noexcept>::value,
+        Variant<Errors...>,
+        deduplicate_t<Variant<Errors..., std::exception_ptr>>>;
+  };
+
+  template <
+      template <typename...> class Variant,
+      template <typename...> class Tuple>
+  using value_types = deduplicate_t<typename Predecessor::template value_types<
+      Variant,
+      indexed_for_result<Tuple>::template apply>>;
+
+  template <template <typename...> class Variant>
+  using error_types = typename Predecessor::template error_types<
+      calculate_errors<Variant>::template apply>;
+
+  friend constexpr auto tag_invoke(
+      tag_t<blocking>,
+      const indexed_for_sender& sender) {
+    return blocking(sender.pred_);
+  }
+
+  template <typename Receiver>
+  struct indexed_for_receiver {
+    UNIFEX_NO_UNIQUE_ADDRESS Func func_;
+    UNIFEX_NO_UNIQUE_ADDRESS Policy policy_;
+    UNIFEX_NO_UNIQUE_ADDRESS Range range_;
+    UNIFEX_NO_UNIQUE_ADDRESS Receiver receiver_;
+
+    // sequenced_policy version supports forward range
+    template<typename... Values>
+    void apply_func_with_policy(const execution::sequenced_policy& policy, Range&& range, Func&& func, Values&... values)
+        noexcept(noexcept(std::invoke((Func &&) func_, std::declval<typename Range::iterator::value_type>(), values...))) {
+      for(auto idx : range) {
+        std::invoke((Func &&) func, idx, values...);
+      }
+    }
+
+    // parallel_policy version requires random access range
+    template<typename... Values>
+    void apply_func_with_policy(const execution::parallel_policy& policy, Range&& range, Func&& func, Values&... values)
+        noexcept(noexcept(std::invoke((Func &&) func_, std::declval<typename Range::iterator::value_type>(), values...))) {
```

Comment:
Should be `func` rather than `func_`?

I think you should be invoking the lvalue-overload of `Func::operator()` since you will be calling it multiple times.

I think you should be using the `iterator::reference` type rather than the `iterator::value_type` here.
Also, prefer to use `std::iterator_traits` to find the members rather than looking for nested type-names.
The nested type-names won't work if the iterator type is a pointer (like it might be for `std::vector`/`std::array`).

You could use `std::is_nothrow_invocable_v<Func&, typename std::iterator_traits<typename Range::iterator>::reference, Values&...>`.

Technically you probably also need to check for noexceptness of:
* `range.size()`
* `iter::operator[]`, `iter::operator++()`

Sometimes I really wish we had a `noexcept(auto)` 😜

Reply:
It should. Making it static forced me to correct that downstream.

Yes, it took me a while to get that noexcept(noexcept( nonsense compiling at all. PR 51 had a rewrite anyway because invoking as above needs this, because of func_ but more generally was messy so I cleaned it up to use the trait.